hasher-matcher-actioner/hmalib/lambdas/api/action_rules_api.py

```diff
@@ -0,0 +1,165 @@
+
+
+@dataclass
+class ActionRulesAPIRequest(DictParseable):
+    action_rule: ActionRule
+
+    @classmethod
+    def from_dict(cls, d: dict) -> "ActionRulesAPIRequest":
+        ar = ActionRule.from_aws(d["action_rule"])
+        logger.info("Deserialized ActionRule: %s", ar)
+        return cls(ar)
+
+
+@dataclass
+class ActionRulesAPIResponse(JSONifiable):
+    error_message: str
+    action_rules: t.List[ActionRule] = field(default_factory=list)
+
+    def to_json(self) -> t.Dict:
+        return {
+            "error_message": self.error_message,
+            "action_rules": [action_rule.to_aws() for action_rule in self.action_rules],
+        }
+
+
+def handle_unexpected_error(e: Exception):
+    logger.error("Unexpected error: %s", sys.exc_info()[0])
+    logger.exception(e)
+    response.status = 500
+
+
+def get_action_rules_api(hma_config_table: str) -> bottle.Bottle:
+    # The documentation below expects prefix to be '/action-rules/'
+    action_rules_api = bottle.Bottle()
+    HMAConfig.initialize(hma_config_table)
+
+    @action_rules_api.get("/", apply=[jsoninator])
+    def get_action_rules() -> ActionRulesAPIResponse:
+        """
+        Returns all action rules.
+        """
+        error_message = ""
+        action_rules = []
+
+        try:
+            action_rules = ActionRule.get_all()
+            logger.info("action_rules: %s", action_rules)
+        except Exception as e:
+            error_message = "Unexpected error."
+            handle_unexpected_error(e)
+
+        return ActionRulesAPIResponse(error_message, action_rules)
+
+    @action_rules_api.post("/", apply=[jsoninator(ActionRulesAPIRequest)])
+    def create_action_rule(
+        request: ActionRulesAPIRequest,
+    ) -> ActionRulesAPIResponse:
+        """
+        Creates an action rule.
+        """
+        logger.info("request: %s", request)
+        error_message = ""
+
+        try:
+            hmaconfig.create_config(request.action_rule)
+        except ClientError as e:
+            # TODO this test for "already exists" should be moved to a common place
+            if e.response["Error"]["Code"] == "ConditionalCheckFailedException":
+                error_message = f"An action rule with the name '{request.action_rule.name}' already exists."
+                logger.warning(
+                    "Duplicate action rule creation attempted: %s",
+                    e.response["Error"]["Message"],
+                )
+            else:
+                error_message = "Unexpected error."
+                logger.error(
+                    "Unexpected client error: %s", e.response["Error"]["Message"]
+                )
+                logger.exception(e)
+            response.status = 500
+        except Exception as e:
+            error_message = "Unexpected error."
+            handle_unexpected_error(e)
+
+        return ActionRulesAPIResponse(error_message)
+
+    @action_rules_api.put("/<old_name>", apply=[jsoninator(ActionRulesAPIRequest)])
```

Comment:
Why not have them both past in as part of the request (or have `oldname` be a query param) and just have the endpoint called `/update/`?

it seems like `/<key>` type endpoints are more trouble than they are worth due to the possibility of '/' in the names and other tbd routing concerns.

Reply:
I'm pretty sure the standard form of a web api for PUTs is to have the ID of the thing you're updating be in the pathing. Check out the first code block of https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/

Reply:
I agree that is it standard, I just know we've had issue with path due to `/` in names etc. and this was a quick way to try and handle that instead. Not worth blocking on, just means the sooner we can add field validations the better.

Reply:
Yeah I'd rather have the fact that names with slashes break the system and push us toward implementing unique IDs be the solution.